056-kafka-roller-kraft.md

```diff
@@ -0,0 +1,129 @@
+ - Does not roll a broker if doing so would take the in-sync replicas count below `min.insync.replicas`
+
+### Configuration changes
+ - Compares the current configuration with expected configuration to see if a broker needs rolling
+ - Performs dynamic configuration updates if possible, otherwise rolls broker on configuration change
+
+In KRaft mode the KafkaRoller currently skips controller only nodes, but performs the above steps on any combined or broker only nodes.
+This is causing a problem in combined mode because if the quorum has not formed due to some of the nodes not being ready 
+the KafkaRoller will wait still try to contact the broker via the admin client.
+This call fails because the quorum is not formed, so in some cases this results in the cluster being stuck with some nodes 
+in a pending state.
+
+## Motivation
+
+When moving to KRaft mode the controller pods need to be rolled if configuration changes occur, also at the moment the 
+existing logic is blocking the full implementation of liveness and readiness checks as described in [proposal #46](https://github.com/strimzi/proposals/blob/main/046-kraft-liveness-readiness.md).
+
+## Proposal
+
+The KafkaRoller behaviour should be unchanged when operating against a ZooKeeper based cluster.
+
+The behaviour when operating against a KRaft cluster is described below.
+
+### Order to roll nodes
+Roll in order:
+- Unready controller/combined nodes
+- Ready controller/combined nodes
+- Active controller
+- Unready broker-only nodes
+- Ready broker-only nodes
+
+If KafkaRoller fails to roll the controller/combined nodes, it still tries to roll the brokers before reporting an error.
+This is required because until [KIP-919](https://cwiki.apache.org/confluence/display/KAFKA/KIP-919%3A+Allow+AdminClient+to+Talk+Directly+with+the+KRaft+Controller+Quorum) 
+is implemented in order to check the quorum state when rolling controllers at least one broker needs to be running.
+
+### Triggers
+The following will trigger an attempt to roll a KRaft controller or combined node:
+- Pod is annotated for manual rolling update
+- Pod is unready
+- Controller's configuration has changed and cannot be updated dynamically.
+
+The triggers for broker remain the same as Zookeeper mode.
+
+**NOTE** There is a related proposal describing how to diff controller configuration currently being discussed: [PR #82](https://github.com/strimzi/proposals/pull/82)
+
+### Rollability checks
+
+The checks made by the KafkaRoller in different modes is described below.
+The checks include a new check for controllers to verify that rolling the node does not affect the quorum health.
+The proposed check is:
+- Fetch `lastCaughtUpTimestamp` for every node
+- A node is marked as caught up if `leaderLastCaughtUpTimestamp - lastCaughtUpTimestamp < controllerQuorumFetchTimeoutMs`
+- Count each node that is caught up, plus the leader (`numOfCaughtUpVoters`)
+- Can roll if: `numOfCaughtUpVoters > (numOfVoters)/2`
```

Comment:
I think properly explaining these values would be useful
* Is number of voters the same as number of controller replicas? 
* What if there is a scale-up / scale-down event? Is it the old or new number of replicas?
* Assuming `numOfVoters` is number of controllers ... would the formula be `numOfCaughtUpVoters > ceil( (double) numOfVoters / 2)`?
* Would it be safer to just go always only by one node? That might allow some non-voluntary disruptions to happen in parallel.
* What role does the Pod readiness play in this?

Reply:
- yes it is, I used `controller` instead of `voters`.
- I believe it would be the current number of replicas as we would be getting this number from the quorum data returned via admin client API. I can try make this a bit clearer. 
- Yes, agree on this. 
- Not sure if I follow this question, could you please elaborate little more? To be honest I think the formula needs little more thought. 
- The quorum would not be checked if the pod is unready. Similar to how it works today with zookeeper based brokers. Their availability does not get checked if pod is unready, instead it gets force rolled. Again I can try make this clearer.

Reply:
> I believe it would be the current number of replicas as we would be getting this number from the quorum data returned via admin client API. I can try make this a bit clearer.

As per some tests I did in the past, it seems that the `describeMetadataQuorum` call doesn't update the list of voters on scale up/scale down. When I raised this, I think @showuon was going to open a KIP to fix this problem. Is it still not fixed? Is it still the case?